File: Striver/DP/Q4.java
package Striver.DP;

public class Q4 {
    // frog jump problem - k jumps allowed

    // recursive solution - Time: O(k^n) Space: O(n)
    public static int frogJump(int n, int k, int[] cost){
        if(n == 0)
            return 0;

        int minVal = Integer.MAX_VALUE;
        for(int i = 1;i<=k;i++){
            if(n-i >= 0){
                int left = frogJump(n-i, k, cost) + Math.abs(cost[n] - cost[n-i]);
                minVal = Math.min(minVal, left);
            }
        }
        return minVal;
    }

    // memoization - Time: O(n*k) Space: O(n)
    public static int frogJumpMemo(int n, int k, int[] cost, int[] dp){
        if(n == 0)
            return 0;
        if(dp[n] != -1)
            return dp[n];

        int minVal = Integer.MAX_VALUE;
        for(int i = 1;i<=k;i++){
            if(n-i >= 0){
                int left = frogJumpMemo(n-i, k, cost, dp) + Math.abs(cost[n] - cost[n-i]);
                minVal = Math.min(minVal, left);
            }
        }
        return dp[n] = minVal;
    }

    // tabulation - Time: O(n*k) Space: O(n)
    public static int frogJumpTab(int n, int k, int[] cost){
        int[] dp = new int[n+1];
        dp[0] = 0;
        for(int i = 1;i<=n;i++){
            int minVal = Integer.MAX_VALUE;
            for(int j = 1;j<=k;j++){
                if(i-j >= 0){
                    int left = dp[i-j] + Math.abs(cost[i] - cost[i-j]);
                    minVal = Math.min(minVal, left);
                }
            }
            dp[i] = minVal;
        }
        return dp[n];
    }

    // space optimization - Time: O(n*k) Space: O(k)
    public static int frogJumpTabOpt(int n, int k, int[] cost){
        int[] dp = new int[k+1];
        dp[0] = 0;
        for(int i = 1;i<=n;i++){
            int minVal = Integer.MAX_VALUE;
            for(int j = 1;j<=k;j++){
                if(i-j >= 0){
                    int left = dp[j] + Math.abs(cost[i] - cost[i-j]);
                    minVal = Math.min(minVal, left);
                }
            }
            dp[i%k] = minVal;
        }
        return dp[n%k];
    }

    public static void main(String[] args) {
        int n = 5;
        int k = 3;
        int[] cost = {0, 10, 30, 40, 50, 60};
        int[] dp = new int[n+1];
        for(int i = 0;i<=n;i++)
            dp[i] = -1;
        System.out.println(frogJump(n, k, cost));
        System.out.println(frogJumpMemo(n, k, cost, dp));
        System.out.println(frogJumpTab(n, k, cost));
        System.out.println(frogJumpTabOpt(n, k, cost));
    }
}

File: Striver/DP/Q4.py
# frog jump problem - k jumps allowed
import sys
max_size = sys.maxsize

# recursive solution
def frogJump(n, k, cost):
    if n == 0:
        return 0
    minVal = max_size
    for i in range(1, k+1):
        if n-i >= 0:
            minVal = min(minVal, frogJump(n-i, k, cost) + abs(cost[n] - cost[n-i]))
    return minVal

# memoization
def frogJump(n, k, cost, dp):
    if n == 0:
        return 0
    if dp[n] != -1:
        return dp[n]
    minVal = max_size
    for i in range(1, k+1):
        minVal = min(minVal, frogJump(n-i, k, cost, dp) + abs(cost[n] - cost[n-i]))
    dp[n] = minVal
    return dp[n]

# tabulation
def frogJump(n,k,cost):
    dp = [-1] * (n+1)
    dp[0] = 0
    for i in range(1,n+1):
        minVal = max_size
        for j in range(1,k+1):
            if i-j >= 0:
                minVal = min(minVal, dp[i-j] + abs(cost[i] - cost[i-j]))
        dp[i] = minVal
    return dp[n]

# space optimazation
def frogJump(n,k,cost):
    dp = [-1] * (k+1)
    dp[0] = 0
    for i in range(1,n+1):
        minVal = max_size
        for j in range(1,k+1):
            if i-j >= 0:
                minVal = min(minVal, dp[j] + abs(cost[i] - cost[i-j]))
        dp[i%k] = minVal
    return dp[n%k]

k = 3
cost = [10, 30, 40, 20]
print(frogJump(len(cost)-1, k, cost))
dp = [-1] * (len(cost))
print(frogJump(len(cost)-1, k, cost, dp))
print(frogJump(len(cost)-1, k, cost))
